File: solutions/python/solution1/src/video_playlist.py
"""A video playlist class."""

class VideoPlaylistError(Exception):
    pass


class VideoPlaylist:
    """A class used to represent a Playlist."""

    def __init__(self, name:str):
        self._name = name
        # Keep the videos as a list
        self._videos = []

    @property
    def name(self):
        return self._name

    @property
    def videos(self):
        return tuple(self._videos)

    def __contains__(self, video):
        """Overloading this method will allow us to use the python "in"
        operator. So now we can do `if video in playlist` like it was a list."""
        return video in self._videos

    def add_video(self, video):
        if video in self:
            raise VideoPlaylistError("Video already added")
        self._videos.append(video)

    def remove_video(self, video):
        if video not in self:
            raise VideoPlaylistError("Video is not in playlist")
        self._videos.remove(video)

    def clear(self):
        self._videos.clear()

    def __str__(self):
        """Overloading __str__ allows us to use print(..) with this object.
                When we do print(playlist) we just want to print the name."""
        return self._name

File: solutions/java/solution3/src/main/java/com/google/VideoPlaylist.java
package com.google;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/** A class used to represent a Playlist */
class VideoPlaylist {

  private final String name;
  private final ArrayList<Video> videos = new ArrayList();
  private Set<String> videoLookup = new HashSet<>();

  String getName() { return name; }

  public VideoPlaylist(String name) {
    this.name = name;
  }

  public void addVideo(Video v) {
    videos.add(v);
    videoLookup.add(v.getVideoId());
  }

  public boolean removeVideo(Video v) {
    if(containsVideo(v.getVideoId())) {
      videos.remove(v);
      videoLookup.remove(v.getVideoId());
      return true;
    } else {
      return false;
    }
  }

  public void clear() {
    videos.clear();
    videoLookup.clear();
  }

  public boolean containsVideo(String videoId) {
    return videoLookup.contains(videoId);
  }

  List<Video> getVideos() {
    return Collections.unmodifiableList(videos);
  }
}